Fixed #35631 -- Extract quality value from Accept header (#18415)
django/django: docs/ref/request-response.txt

    Returns ``True`` if the request is secure; that is, if it was made with
    HTTPS.

.. method:: HttpRequest.get_preferred_type(media_types)

    .. versionadded:: 5.2

>     Returns the most preferred mime type from ``media_types`` which the client
>     accepts, according to the ``Accept`` header.

Reviewer: ```suggestion
    Returns the preferred mime type from ``media_types``, based on the
    ``Accept`` header.
```

    Assuming the client sends an ``Accept`` header of
    ``text/html,application/json;q=0.8``:

    .. code-block:: pycon

        >>> request.get_preferred_type(["text/html", "application/json"])
        "text/html"
        >>> request.get_preferred_type(["application/json", "text/plain"])

>         "application/json"

Reviewer: Reading this I'm left wondering, so could you add an example for a non-matching list, so as... 

    >>> request.get_preferred_type(["application/xml", "text/plain"])
    None

(Probably worth a sentence.)

Users can request what they like. I've got what I've got 😅

    Returns ``True`` if the request is secure; that is, if it was made with
    HTTPS.

.. method:: HttpRequest.get_preferred_type(media_types)

    .. versionadded:: 5.2

    Returns the preferred mime type from ``media_types`` based on to the
    ``Accept`` header, or ``None`` if the client does not accept any of the

>     provided types.

Reviewer: `s/based on to the/based on the/`

    Assuming the client sends an ``Accept`` header of
    ``text/html,application/json;q=0.8``:

    .. code-block:: pycon

        >>> request.get_preferred_type(["text/html", "application/json"])
        "text/html"
        >>> request.get_preferred_type(["application/json", "text/plain"])
        "application/json"
        >>> request.get_preferred_type(["application/xml", "text/plain"])
        None

>     See :ref:`content-negotiation-example` for an example of using

Reviewer 2: The diff here is a bit odd, I think we've moved this section out from under `HttpRequest.accepts(mime_type)`:
```See :ref:`content-negotiation-example` for an example of using ``accepts()`` ```
I'm not sure that's what was intended?

Perhaps you wanted a new link to ```:ref:`content-negotiation-example` ``` for the example of using `get_preferred_type()`?

Author: I've fixed the diff a little, so both `accepts` and `get_preferred_type` link to the example. My main intention was to make `get_preferred_type` the main API by moving most of the documentation onto it.

    Returns ``True`` if the request is secure; that is, if it was made with
    HTTPS.

.. method:: HttpRequest.get_preferred_type(media_types)

    .. versionadded:: 5.2

    Returns the preferred mime type from ``media_types``, based on the
    ``Accept`` header, or ``None`` if the client does not accept any of the
    provided types.

    Assuming the client sends an ``Accept`` header of
    ``text/html,application/json;q=0.8``:

    .. code-block:: pycon

        >>> request.get_preferred_type(["text/html", "application/json"])
        "text/html"
        >>> request.get_preferred_type(["application/json", "text/plain"])
        "application/json"
        >>> request.get_preferred_type(["application/xml", "text/plain"])
        None

    Most browsers send ``Accept: */*`` by default, so this would return

>     ``True`` for all content types. Setting an explicit ``Accept`` header in

Reviewer 2: I still think this is mixing up elements from `HttpRequest.accepts(mime_type)`, so example this says:

```Most browsers send ``Accept: */*`` by default, so this would return ``True`` for all content types.```

But `accepts` would return `True`, this would return the first item of `mime_types`

I think this sentence should be adapted for `HttpRequest.get_preferred_type(media_types)` and this sentence also added back in to `HttpRequest.accepts(mime_type)`

I'm wondering for we should best show that the ```If a response varies depending on the content of the ``Accept`` header...``` paragraph applies to both of these and is more general advice - maybe we should make this a note or admonition or something :thinking:

Author: I've tweaked the wording a little, and extracted the caching point into a `note`. Because each attribute is just listed, without grouping, it's hard to find a common place, but at least putting in on the first (and preferred) API should make it easier to see.
